hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java

```diff
@@ -197,6 +197,27 @@
      */
     private final Runnable taskDoneCallback;
 
+    /**
+     * Shows whether invocation can retry or not. It has two purposes.
+     *
+     * <ul><li>
+     * Control access to {@link Invocation#doInvoke(boolean)}. This method
+     * can be accessed from the invoker thread or the single threaded
+     * scheduler of {@link InvocationMonitor}. This variable ensures that
+     * when scheduler retries, invoker thread will no longer read/write to
+     * a shared non-final variable.
+     * </li><li>
+     * Provide a happens-before relationship between invoker thread,
+     * scheduler and callers of {@link Invocation#toString()}.
+     * </li></ul>
+     *
+     * <p>
+     * Note that after the invocation is registered, this variable must be
+     * set to {@code true} eventually. Also make sure that there is no read
+     * or write to a shared variable after it's set.
+     */
+    private volatile boolean invocationCanRetry;
```

Comment:
instead of throwing exception, what if we return `false` here? Can it be a fix also?

https://github.com/hazelcast/hazelcast/blob/f35a6f1e3bf77fdf70fd4d542c40f7b30746a330/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OutboundOperationHandler.java#L55-L57

Reply:
It's enough to not fail because of `IllegalArgumentException` but not enough to solve the issue. Note that there is no synchronization between invoker and monitor thread. This means that monitor thread even can see wrong target. Also there is a problem this PR solves about `OnMemberLeftTask`.

Reply:
It seems I was wrong, this is enough for the first issue solved in this PR. For the issue related to `OnMemberLeftTask`, I will investigate more.